File: java/leetcode/RussianDoll.java
//Leetcode problem 354 Russian Doll Envelopes
//Solution written by Xuqiang Fang on 19 June, 2018
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.HashSet;
import java.util.Set;
import java.util.PriorityQueue;
import java.util.Stack;
import java.util.Queue;
import java.util.LinkedList;
import java.util.Arrays;
import java.util.Comparator;
class Solution{
    public int maxEnvelopes(int[][] envelopes) {
        int n = envelopes.length;
        if(n == 0){
            return 0;
        }
        Arrays.sort(envelopes, new MyCom());
        int[] dp = new int[n];
        Arrays.fill(dp, 1);
        int ans = 1;
        for(int i=1; i<n; ++i){
            for(int j=0; j<i; ++j){
                if(envelopes[i][0] > envelopes[j][0] && envelopes[i][1] > envelopes[j][1]){
                    dp[i] = Math.max(dp[i], dp[j]+1);
                }
            }
            ans = Math.max(ans, dp[i]);
        }
        return ans;
    }
}
class MyCom implements Comparator<int[]>{
    public int compare(int[] a, int[] b){
        if(a[0] == b[0]){
            return a[1] - b[1];
        } 
        return a[0] - b[0];
    }
}

public class RussianDoll{
	public static void main(String[] args){
		Solution s = new Solution();
        int[][] envelopes = {{5,4},{6,4},{6,7},{2,3}};
        System.out.println(s.maxEnvelopes(envelopes));
	}
}

File: python/leetcode/RussianDoll.py
'''
Leetcode problem No 354 Russian Doll Envelopes
Solution written by Xuqiang Fang on 19 June, 2018
'''

class Solution(object):
    #### this solution is TLE
    def maxEnvelopes(self, en):
        """
        :type envelopes: List[List[int]]
        :rtype: int
        """
        n = len(en)
        if n == 0:
            return 0
        en.sort()
        dp = [1 for i in range(n)]
        ans = 1
        for i in range(1, n):
            for j in range(0, i):
                if en[i][1] > en[j][1] and en[i][0] > en[j][0]:
                    dp[i] = max(dp[i], dp[j]+1)
            ans = max(ans, dp[i])
        return ans

def main():
    s = Solution()
    envelopes = [[5,4],[6,4],[6,7],[2,3]]
    envelop = [[4,5],[4,6],[6,7],[2,3],[1,1]]
    print(s.maxEnvelopes(envelopes))
    print(s.maxEnvelopes(envelop))
main()
